Optimizing K Framework Part 2/article.md

```diff
@@ -0,0 +1,552 @@
+
+```hs
+data PreHashed a = PreHashed
+    { preHashed :: !a
+    , preHash :: ~Int
+    }
+    deriving stock (Generic)
+    deriving anyclass (NFData)
+
+mkPreHashed :: Hashable a => a -> PreHashed a
+mkPreHashed !a = PreHashed a (hash a)
+{-# INLINE mkPreHashed #-}
+```
+This type is similar to the `Data.Hashable.Hashed` type, but it has a few differences:
+* It has a lazy hash.
+* It has a strict `a`.
+* The `Eq` instance does not use the hash (to avoid evaluating it).
+
+This change, however, resulted in a negligible impact with only a 0.33% overall speedup.
+
+Our [next attempt](https://github.com/runtimeverification/haskell-backend/commits/jb/caching-termlike-hashes) was to pre-compute hash for the term and store it in `TermAttributes` of `TermLike`. This pre-computed hash is then accessed by the Hashable instance rather than recursively traversing the `TermLikeF` data structure. To implement this, we added a new field `termHash` to `TermAttributes` to store the pre-computed hash value, just like we did for the `Equation` type:
+
+```hs
+data TermAttributes variable = TermAttributes
+    { termSort :: !Sort
+    , termFreeVariables :: !(Attribute.FreeVariables variable)
+    , termFunctional :: !Attribute.Functional
+    , termFunction :: !Attribute.Function
+    , termDefined :: !Attribute.Defined
+    , termCreated :: !Attribute.Created
+    , termSimplified :: !Attribute.Simplified
+    , termConstructorLike :: !Attribute.ConstructorLike
+    , termHash :: Int -- caches the TermLikeF hash value for efficiency
+    }
+```
+
+To compute the hash value using the `hashWithSalt` function, we use the pre-computed hash value stored in `termHash`, along with the provided `salt` value:
+
+```hs
+instance Hashable variable => Hashable (TermLike variable) where
+    hashWithSalt salt (Recursive.project -> TermAttributes{termHash} :< _) =
+        salt `hashWithSalt` termHash
+    {-# INLINE hashWithSalt #-}
+
+```
+
+To produce the pre-computed hash value, we use the derived `Hashable` instance of `TermLikeF`, which will call hash on the respective sub-structure wrapped into one of the `TermLikeF` constructors. For all such sub-structures, the derived `Hashable` instance will then recurse into the contents, using the cached hashes directly instead of performing a deep recursion.
+
+Although the change resulted in a 6% speedup, it also caused some integration tests to fail, and unfortunately, we were unable to find a way to fix them, while keeping pre-computed hash in `TermAttributes`.
+So, to avoid storing pre-computed hash in `TermAttributes`, we [cached the hash](https://github.com/runtimeverification/haskell-backend/pull/3338) of the `TermLikeF` in the `TermLike` itself and extract it in the `Hashable` instance instead of recursively traversing into the `TermLikeF`. To achieve this, we modified the `TermLike` definition:
```

Comment:
```suggestion
So, to avoid storing pre-computed hash in `TermAttributes`, we [cached the hash](https://github.com/runtimeverification/haskell-backend/pull/3338) of the `TermLikeF` in the `TermLike` itself and extracted it in the `Hashable` instance instead of recursively traversing into the `TermLikeF`. To achieve this, we modified the `TermLike` definition:
```